cms_form_compassion/forms/payment_form.py

```diff
@@ -7,125 +7,62 @@
-
-        def _form_validate_amount(self, value, **req_values):
-            try:
-                amount = float(value)
-                if amount <= 0:
-                    raise ValueError
-            except (ValueError, TypeError):
-                return "amount", _("Please control the amount")
-            # No error
-            return 0, 0
-
-        def form_next_url(self, main_object=None):
-            # Redirect to payment controller, creating a transaction
-            return "/compassion/payment/{}?redirect_url={}&display_type={}".format(
-                str(self.transaction_id),
-                self._payment_accept_redirect,
-                self._display_type,
-            )
-
-        def form_after_create_or_update(self, values, extra_values):
-            """ Dismiss status message, as the client will be redirected
-            to payment."""
-            super().form_after_create_or_update(values, extra_values)
-            self.o_request.website.get_status_message()
-            all_vals = self.get_all_vals(values, extra_values)
-            partner = (
-                self.env["res.partner"]
-                .sudo()
-                .browse(all_vals.get("partner_id"))
-                .exists()
-            )
-            tx_values = {
-                "acquirer_id": all_vals.get("acquirer_id"),
-                "type": "form",
-                "amount": all_vals.get("amount"),
-                "currency_id": all_vals.get("currency_id"),
-                "partner_id": partner.id,
-                "partner_country_id": partner.country_id.id,
-                "reference": "payment-form",
-            }
-            self._edit_transaction_values(tx_values, all_vals)
-            transaction_obj = self.env["payment.transaction"].sudo()
-            tx_values["reference"] = transaction_obj.get_next_reference(
-                tx_values["reference"]
-            )
-            self.transaction_id = transaction_obj.create(tx_values).id
-            self.request.session["compassion_transaction_id"] = self.transaction_id
-
-        def get_all_vals(self, values, extra_values):
-            """
-            Used to find form value that can either be in values or
-            extra_values dictionary.
-            :param values: values dictionary (containing form model fields)
-            :param extra_values: extra_values dictionary
-                                 (containing non-model fields)
-            :return: new dictionary that intersect the two sources
-            """
-            all_vals = values.copy()
-            all_vals.update(extra_values)
-            return all_vals
 
-        def _edit_transaction_values(self, tx_values, form_vals):
-            """ Hook to setup custom values for payment. """
-            pass
+    def generate_invoice(self):
+        """Inherit this method in order to create the invoice that will
+        be used for the web payment."""
+        return self.env["account.invoice"]
```

Comment:
is this correct ? Shouldn't it be self.env["account.invoice"].create() ?

Reply:
This is the abstract form which purpose is to be inherited. Returning this will just return an empty recordset (using create would raise an error because you cannot create an invoice without any values).